File: mobile/lib/services/state_builder.dart
import '../models/contact.dart';
import '../models/transaction.dart';
import '../models/event.dart';
import 'event_store_service.dart';

/// Application state containing all contacts and transactions
class AppState {
  final List<Contact> contacts;
  final List<Transaction> transactions;
  final DateTime lastBuiltAt;

  AppState({
    required this.contacts,
    required this.transactions,
    required this.lastBuiltAt,
  });

  AppState copyWith({
    List<Contact>? contacts,
    List<Transaction>? transactions,
    DateTime? lastBuiltAt,
  }) {
    return AppState(
      contacts: contacts ?? this.contacts,
      transactions: transactions ?? this.transactions,
      lastBuiltAt: lastBuiltAt ?? this.lastBuiltAt,
    );
  }
}

/// StateBuilder - Simple, pure functions to build state from events
/// KISS principle: No side effects, easy to test, incremental updates
class StateBuilder {
  /// Build full state from all events
  static AppState buildState(List<Event> events) {
    // Sort events by timestamp to ensure correct order
    final sortedEvents = List<Event>.from(events)
      ..sort((a, b) => a.timestamp.compareTo(b.timestamp));

    final contacts = <String, Contact>{};
    final transactions = <String, Transaction>{};

    for (final event in sortedEvents) {
      if (event.aggregateType == 'contact') {
        _applyContactEvent(contacts, event);
      } else if (event.aggregateType == 'transaction') {
        _applyTransactionEvent(transactions, event, contacts);
      }
    }

    // Calculate balances for all contacts
    _calculateBalances(contacts, transactions.values.toList());

    return AppState(
      contacts: contacts.values.toList(),
      transactions: transactions.values.toList(),
      lastBuiltAt: DateTime.now(),
    );
  }

  /// Apply new events to existing state (incremental update)
  static AppState applyEvents(AppState currentState, List<Event> newEvents) {
    if (newEvents.isEmpty) return currentState;

    // Sort new events by timestamp
    final sortedNewEvents = List<Event>.from(newEvents)
      ..sort((a, b) => a.timestamp.compareTo(b.timestamp));

    // Convert current state to maps for easier manipulation
    final contacts = Map<String, Contact>.fromEntries(
      currentState.contacts.map((c) => MapEntry(c.id, c)),
    );
    final transactions = Map<String, Transaction>.fromEntries(
      currentState.transactions.map((t) => MapEntry(t.id, t)),
    );

    // Apply new events
    for (final event in sortedNewEvents) {
      if (event.aggregateType == 'contact') {
        _applyContactEvent(contacts, event);
      } else if (event.aggregateType == 'transaction') {
        _applyTransactionEvent(transactions, event, contacts);
      }
    }

    // Recalculate balances (only affected contacts need recalculation)
    _calculateBalances(contacts, transactions.values.toList());

    return AppState(
      contacts: contacts.values.toList(),
      transactions: transactions.values.toList(),
      lastBuiltAt: DateTime.now(),
    );
  }

  /// Apply a contact event to the contacts map
  static void _applyContactEvent(Map<String, Contact> contacts, Event event) {
    final contactId = event.aggregateId;
    final eventData = event.eventData;

    if (event.eventType == 'CREATED') {
      contacts[contactId] = Contact(
        id: contactId,
        name: eventData['name'] as String? ?? '',
        username: eventData['username'] as String?,
        phone: eventData['phone'] as String?,
        email: eventData['email'] as String?,
        notes: eventData['notes'] as String?,
        createdAt: _parseTimestamp(eventData['timestamp']),
        updatedAt: _parseTimestamp(eventData['timestamp']),
        isSynced: event.synced,
        balance: 0, // Will be calculated later
      );
    } else if (event.eventType == 'UPDATED' && contacts.containsKey(contactId)) {
      final existing = contacts[contactId]!;
      contacts[contactId] = Contact(
        id: existing.id,
        name: eventData['name'] as String? ?? existing.name,
        username: eventData['username'] as String? ?? existing.username,
        phone: eventData['phone'] as String? ?? existing.phone,
        email: eventData['email'] as String? ?? existing.email,
        notes: eventData['notes'] as String? ?? existing.notes,
        createdAt: existing.createdAt,
        updatedAt: _parseTimestamp(eventData['timestamp']),
        isSynced: event.synced,
        balance: existing.balance, // Will be recalculated
      );
    } else if (event.eventType == 'DELETED' && contacts.containsKey(contactId)) {
      // Remove deleted contact (hard delete for simplicity)
      contacts.remove(contactId);
    }
  }

  /// Apply a transaction event to the transactions map
  static void _applyTransactionEvent(
    Map<String, Transaction> transactions,
    Event event,
    Map<String, Contact> contacts,
  ) {
    final transactionId = event.aggregateId;
    final eventData = event.eventData;

    if (event.eventType == 'CREATED') {
      final contactId = eventData['contact_id'] as String? ?? '';
      // Only create if contact exists
      if (contacts.containsKey(contactId)) {
        transactions[transactionId] = Transaction(
          id: transactionId,
          contactId: contactId,
          type: eventData['type'] == 'item'
              ? TransactionType.item
              : TransactionType.money,
          direction: eventData['direction'] == 'owed'
              ? TransactionDirection.owed
              : TransactionDirection.lent,
          amount: (eventData['amount'] as num?)?.toInt() ?? 0,
          currency: eventData['currency'] as String? ?? 'IQD',
          description: eventData['description'] as String?,
          transactionDate: _parseDate(eventData['transaction_date']) ?? DateTime.now(),
          dueDate: _parseDate(eventData['due_date']),
          createdAt: _parseTimestamp(eventData['timestamp']),
          updatedAt: _parseTimestamp(eventData['timestamp']),
          isSynced: event.synced,
        );
      }
    } else if (event.eventType == 'UPDATED' && transactions.containsKey(transactionId)) {
      final existing = transactions[transactionId]!;
      transactions[transactionId] = Transaction(
        id: existing.id,
        contactId: eventData['contact_id'] as String? ?? existing.contactId,
        type: eventData['type'] == 'item'
            ? TransactionType.item
            : (eventData['type'] == 'money' ? TransactionType.money : existing.type),
        direction: eventData['direction'] == 'owed'
            ? TransactionDirection.owed
            : (eventData['direction'] == 'lent'
                ? TransactionDirection.lent
                : existing.direction),
        amount: (eventData['amount'] as num?)?.toInt() ?? existing.amount,
        currency: eventData['currency'] as String? ?? existing.currency,
        description: eventData['description'] as String? ?? existing.description,
        transactionDate: _parseDate(eventData['transaction_date']) ?? existing.transactionDate,
        imagePaths: existing.imagePaths,
        dueDate: _parseDate(eventData['due_date']) ?? existing.dueDate,
        createdAt: existing.createdAt,
        updatedAt: _parseTimestamp(eventData['timestamp']),
        isSynced: event.synced,
      );
    } else if (event.eventType == 'DELETED' && transactions.containsKey(transactionId)) {
      // Remove transaction (hard delete for simplicity)
      transactions.remove(transactionId);
    }
  }

  /// Calculate balances for all contacts based on transactions
  static void _calculateBalances(
    Map<String, Contact> contacts,
    List<Transaction> transactions,
  ) {
    // Reset all balances
    for (final contact in contacts.values) {
      contacts[contact.id] = Contact(
        id: contact.id,
        name: contact.name,
        username: contact.username,
        phone: contact.phone,
        email: contact.email,
        notes: contact.notes,
        createdAt: contact.createdAt,
        updatedAt: contact.updatedAt,
        isSynced: contact.isSynced,
        balance: 0,
      );
    }

    // Calculate balance from transactions
    for (final transaction in transactions) {
      final contact = contacts[transaction.contactId];
      if (contact != null) {
        final amount = transaction.direction == TransactionDirection.lent
            ? transaction.amount
            : -transaction.amount;
        contacts[transaction.contactId] = Contact(
          id: contact.id,
          name: contact.name,
          username: contact.username,
          phone: contact.phone,
          email: contact.email,
          notes: contact.notes,
          createdAt: contact.createdAt,
          updatedAt: contact.updatedAt,
          isSynced: contact.isSynced,
          balance: contact.balance + amount,
        );
      }
    }
  }

  /// Parse timestamp from event data
  static DateTime _parseTimestamp(dynamic timestamp) {
    if (timestamp == null) return DateTime.now();
    if (timestamp is String) {
      try {
        return DateTime.parse(timestamp);
      } catch (e) {
        return DateTime.now();
      }
    }
    return DateTime.now();
  }

  /// Parse date from event data (YYYY-MM-DD format)
  static DateTime? _parseDate(dynamic date) {
    if (date == null) return null;
    if (date is String) {
      try {
        return DateTime.parse(date);
      } catch (e) {
        return null;
      }
    }
    return null;
  }

  /// Calculate total debt at a specific timestamp
  /// Returns the sum of all contact balances up to that point in time
  static Future<int> calculateTotalDebtAtTime(DateTime timestamp) async {
    // Get all events up to the timestamp
    final allEvents = await EventStoreService.getAllEvents();
    final eventsUpToTime = allEvents
        .where((e) => e.timestamp.isBefore(timestamp) || e.timestamp.isAtSameMomentAs(timestamp))
        .toList();

    // Build state from events up to that time
    final state = buildState(eventsUpToTime);

    // Calculate total debt (sum of all contact balances)
    return state.contacts.fold<int>(0, (sum, contact) => sum + contact.balance);
  }
}

File: frontend/src/state_builder.rs
//! Pure state rebuild from events - ported from Flutter StateBuilder.
//! No side effects, easy to test.

use crate::models::{Contact, Event, Transaction, TransactionDirection, TransactionType};
use chrono::{DateTime, Utc};
use std::collections::HashMap;

/// Application state: contacts and transactions built from events.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct AppState {
    pub contacts: Vec<Contact>,
    pub transactions: Vec<Transaction>,
    pub last_built_at: DateTime<Utc>,
}

/// Build full state from all events (sorted by timestamp).
pub fn build_state(events: &[Event]) -> AppState {
    let mut sorted: Vec<&Event> = events.iter().collect();
    sorted.sort_by(|a, b| a.timestamp.cmp(&b.timestamp));

    let undone_event_ids: std::collections::HashSet<_> = sorted
        .iter()
        .filter(|e| e.event_type == "UNDO")
        .filter_map(|e| e.event_data.get("undone_event_id").and_then(|v| v.as_str()))
        .collect();

    let mut contacts: HashMap<String, Contact> = HashMap::new();
    let mut transactions: HashMap<String, Transaction> = HashMap::new();

    for event in sorted {
        if event.event_type == "UNDO" {
            continue;
        }
        if undone_event_ids.contains(event.id.as_str()) {
            continue;
        }

        if event.aggregate_type == "contact" {
            apply_contact_event(&mut contacts, event, &mut transactions);
        } else if event.aggregate_type == "transaction" {
            apply_transaction_event(&mut transactions, event, &contacts);
        }
    }

    calculate_balances(&mut contacts, transactions.values().cloned().collect());

    AppState {
        contacts: contacts.into_values().collect(),
        transactions: transactions.into_values().collect(),
        last_built_at: Utc::now(),
    }
}

/// Apply new events to existing state (incremental update).
pub fn apply_events(current: &AppState, new_events: &[Event]) -> AppState {
    if new_events.is_empty() {
        return current.clone();
    }

    let mut sorted: Vec<&Event> = new_events.iter().collect();
    sorted.sort_by(|a, b| a.timestamp.cmp(&b.timestamp));

    let undone_event_ids: std::collections::HashSet<_> = sorted
        .iter()
        .filter(|e| e.event_type == "UNDO")
        .filter_map(|e| e.event_data.get("undone_event_id").and_then(|v| v.as_str()))
        .collect();

    let mut contacts: HashMap<String, Contact> = current
        .contacts
        .iter()
        .map(|c| (c.id.clone(), c.clone()))
        .collect();
    let mut transactions: HashMap<String, Transaction> = current
        .transactions
        .iter()
        .map(|t| (t.id.clone(), t.clone()))
        .collect();

    for event in sorted {
        if event.event_type == "UNDO" {
            continue;
        }
        if undone_event_ids.contains(event.id.as_str()) {
            continue;
        }

        if event.aggregate_type == "contact" {
            apply_contact_event(&mut contacts, event, &mut transactions);
        } else if event.aggregate_type == "transaction" {
            apply_transaction_event(&mut transactions, event, &contacts);
        }
    }

    calculate_balances(&mut contacts, transactions.values().cloned().collect());

    AppState {
        contacts: contacts.into_values().collect(),
        transactions: transactions.into_values().collect(),
        last_built_at: Utc::now(),
    }
}

fn apply_contact_event(
    contacts: &mut HashMap<String, Contact>,
    event: &Event,
    transactions: &mut HashMap<String, Transaction>,
) {
    let contact_id = event.aggregate_id.clone();
    let data = &event.event_data;

    if event.event_type == "CREATED" {
        let ts = parse_timestamp(data.get("timestamp"));
        contacts.insert(
            contact_id.clone(),
            Contact {
                id: contact_id,
                name: data.get("name").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                username: data.get("username").and_then(|v| v.as_str()).map(String::from),
                phone: data.get("phone").and_then(|v| v.as_str()).map(String::from),
                email: data.get("email").and_then(|v| v.as_str()).map(String::from),
                notes: data.get("notes").and_then(|v| v.as_str()).map(String::from),
                created_at: ts,
                updated_at: ts,
                is_synced: event.synced,
                balance: 0,
                wallet_id: data.get("wallet_id").and_then(|v| v.as_str()).map(String::from),
            },
        );
    } else if event.event_type == "UPDATED" {
        if let Some(existing) = contacts.get(&contact_id) {
            let ts = parse_timestamp(data.get("timestamp"));
            contacts.insert(
                contact_id.clone(),
                Contact {
                    name: data.get("name").and_then(|v| v.as_str()).unwrap_or(&existing.name).to_string(),
                    username: data.get("username").and_then(|v| v.as_str()).map(String::from).or(existing.username.clone()),
                    phone: data.get("phone").and_then(|v| v.as_str()).map(String::from).or(existing.phone.clone()),
                    email: data.get("email").and_then(|v| v.as_str()).map(String::from).or(existing.email.clone()),
                    notes: data.get("notes").and_then(|v| v.as_str()).map(String::from).or(existing.notes.clone()),
                    updated_at: ts,
                    wallet_id: data.get("wallet_id").and_then(|v| v.as_str()).map(String::from).or(existing.wallet_id.clone()),
                    ..existing.clone()
                },
            );
        }
    } else if event.event_type == "DELETED" {
        contacts.remove(&contact_id);
        let to_remove: Vec<String> = transactions
            .values()
            .filter(|t| t.contact_id == contact_id)
            .map(|t| t.id.clone())
            .collect();
        for id in to_remove {
            transactions.remove(&id);
        }
    }
}

fn apply_transaction_event(
    transactions: &mut HashMap<String, Transaction>,
    event: &Event,
    contacts: &HashMap<String, Contact>,
) {
    let txn_id = event.aggregate_id.clone();
    let data = &event.event_data;

    if event.event_type == "CREATED" {
        let contact_id = data.get("contact_id").and_then(|v| v.as_str()).unwrap_or("").to_string();
        if contact_id.is_empty() || !contacts.contains_key(&contact_id) {
            return;
        }
        let type_ = match data.get("type").and_then(|v| v.as_str()).unwrap_or("money") {
            "item" => TransactionType::Item,
            _ => TransactionType::Money,
        };
        let direction = match data.get("direction").and_then(|v| v.as_str()).unwrap_or("owed") {
            "lent" => TransactionDirection::Lent,
            _ => TransactionDirection::Owed,
        };
        let amount = data.get("amount").and_then(|v| v.as_i64()).unwrap_or(0);
        let currency = data.get("currency").and_then(|v| v.as_str()).unwrap_or("IQD").to_string();
        let transaction_date = parse_date(data.get("transaction_date")).unwrap_or_else(|| chrono::Utc::now().date_naive());
        let created_ts = parse_timestamp(data.get("timestamp"));

        transactions.insert(
            txn_id.clone(),
            Transaction {
                id: txn_id,
                contact_id: contact_id.clone(),
                type_,
                direction,
                amount,
                currency,
                description: data.get("description").and_then(|v| v.as_str()).map(String::from),
                transaction_date,
                due_date: parse_date(data.get("due_date")),
                image_paths: vec![],
                created_at: created_ts,
                updated_at: created_ts,
                is_synced: event.synced,
                wallet_id: data.get("wallet_id").and_then(|v| v.as_str()).map(String::from),
            },
        );
    } else if event.event_type == "UPDATED" {
        if let Some(existing) = transactions.get(&txn_id) {
            let type_ = match data.get("type").and_then(|v| v.as_str()) {
                Some("item") => TransactionType::Item,
                Some("money") => TransactionType::Money,
                _ => existing.type_,
            };
            let direction = match data.get("direction").and_then(|v| v.as_str()) {
                Some("lent") => TransactionDirection::Lent,
                Some("owed") => TransactionDirection::Owed,
                _ => existing.direction,
            };
            let updated_ts = parse_timestamp(data.get("timestamp"));

            transactions.insert(
                txn_id.clone(),
                Transaction {
                    contact_id: data.get("contact_id").and_then(|v| v.as_str()).map(String::from).unwrap_or_else(|| existing.contact_id.clone()),
                    type_,
                    direction,
                    amount: data.get("amount").and_then(|v| v.as_i64()).unwrap_or(existing.amount),
                    currency: data.get("currency").and_then(|v| v.as_str()).map(String::from).unwrap_or_else(|| existing.currency.clone()),
                    description: data.get("description").and_then(|v| v.as_str()).map(String::from).or(existing.description.clone()),
                    transaction_date: parse_date(data.get("transaction_date")).unwrap_or(existing.transaction_date),
                    due_date: parse_date(data.get("due_date")).or(existing.due_date),
                    updated_at: updated_ts,
                    is_synced: event.synced,
                    wallet_id: data.get("wallet_id").and_then(|v| v.as_str()).map(String::from).or(existing.wallet_id.clone()),
                    id: existing.id.clone(),
                    created_at: existing.created_at,
                    image_paths: existing.image_paths.clone(),
                },
            );
        }
    } else if event.event_type == "DELETED" {
        transactions.remove(&txn_id);
    }
}

fn calculate_balances(contacts: &mut HashMap<String, Contact>, transactions: Vec<Transaction>) {
    for c in contacts.values_mut() {
        c.balance = 0;
    }
    for t in transactions {
        if let Some(c) = contacts.get_mut(&t.contact_id) {
            let amount = match t.direction {
                TransactionDirection::Lent => t.amount,
                TransactionDirection::Owed => -t.amount,
            };
            c.balance += amount;
        }
    }
}

fn parse_timestamp(v: Option<&serde_json::Value>) -> DateTime<Utc> {
    let s = match v {
        Some(serde_json::Value::String(s)) => s,
        _ => return Utc::now(),
    };
    chrono::DateTime::parse_from_rfc3339(s)
        .map(|dt| dt.with_timezone(&Utc))
        .unwrap_or_else(|_| Utc::now())
}

fn parse_date(v: Option<&serde_json::Value>) -> Option<chrono::NaiveDate> {
    let s = match v {
        Some(serde_json::Value::String(s)) => s.as_str(),
        _ => return None,
    };
    let s = if s.contains('T') || s.contains(' ') {
        s.to_string()
    } else {
        format!("{}T00:00:00", s)
    };
    chrono::NaiveDateTime::parse_from_str(&s, "%Y-%m-%dT%H:%M:%S")
        .ok()
        .map(|dt| dt.date())
        .or_else(|| chrono::NaiveDate::parse_from_str(s.trim_matches(|c| c == 'T' || c == ' '), "%Y-%m-%d").ok())
}
